### PR TITLE
chore: add workflows, scripts and documentation for release packages

### PR DESCRIPTION
**Pull request by phantinuss**
### Summary of the Pull Request
This PR will add the stuff we need to have automated releases. Will be semi-automatically to begin with (needs to set tags manually which I will do in the beginning). But we can automate it using a cron schedule when things are confirmed stable

To see the generated release page see my fork: https://github.com/phantinuss/sigma/releases

### Changelog
chore: add workflows, scripts and documentation for release packages

### SigmaHQ Rule Creation Conventions

- If your PR adds new rules, please consider following and applying these [conventions](https://github.com/SigmaHQ/sigma-specification/blob/main/sigmahq/sigmahq_conventions.md)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+          echo "#### New rules:" > changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -E '^\s*new: ' | sort | sed -e 's%^% - %' >> changes.txt
+          echo "#### Changed rules:" >> changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -E '^\s*change: ' | sort | sed -e 's%^% - %' >> changes.txt
+          echo "#### Rules with fixes:" >> changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -E '^\s*fix: ' | sort | sed -e 's%^% - %' >> changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP 'Merge PR #\d+ from \K(@\S+)' | sort -u > authors_raw.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP "Co-authored-by: \K.*(?= <)" | sort -u | sed -e 's%^%@%' >> authors_raw.txt
+          LC_ALL=en_US.UTF-8 sort -u authors_raw.txt | grep -v 'dependabot\[bot\]' > authors.txt
+          echo "### Changes in this release:" > changelog.txt
```
Comment (nasbench) [inline .github/workflows/release.yml:32]: No need to mention it imo as the other titles already do a good job at describing what's up 
```suggestion
```

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 0
+      - name: Generate Changelog
+        run: |
+          prev_tag=$(git for-each-ref --sort=creatordate --format '%(refname:lstrip=2)' refs/tags | grep ^r | tail -2 | head -1)
+          curr_tag=$(git for-each-ref --sort=creatordate --format '%(refname:lstrip=2)' refs/tags | grep ^r | tail -1)
+          echo "Previous tag: ${prev_tag}"
+          echo "Current tag: ${curr_tag}"
+          echo "#### New rules:" > changes.txt
```
Comment (nasbench) [inline .github/workflows/release.yml:23]:
```suggestion
          echo "### New Rules" > changes.txt
```

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP 'Merge PR #\d+ from \K(@\S+)' | sort -u > authors_raw.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP "Co-authored-by: \K.*(?= <)" | sort -u | sed -e 's%^%@%' >> authors_raw.txt
+          LC_ALL=en_US.UTF-8 sort -u authors_raw.txt | grep -v 'dependabot\[bot\]' > authors.txt
+          echo "### Changes in this release:" > changelog.txt
+          cat changes.txt >> changelog.txt
+          echo "" >> changelog.txt
+          echo "Thanks to $(perl -pe 's%\n%, %' authors.txt | sed 's%, $%%')" >> changelog.txt
+          echo "" >> changelog.txt
+          echo "" >> changelog.txt
+          echo "#### Which Sigma rule package should I use?" >> changelog.txt
```
Comment (nasbench) [inline .github/workflows/release.yml:38]:
```suggestion
          echo "### Which Sigma rule package should I use?" >> changelog.txt
```

File: .github/workflows/release.yml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -E '^\s*change: ' | sort | sed -e 's%^% - %' >> changes.txt
+          echo "#### Rules with fixes:" >> changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -E '^\s*fix: ' | sort | sed -e 's%^% - %' >> changes.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP 'Merge PR #\d+ from \K(@\S+)' | sort -u > authors_raw.txt
+          git log --pretty=%B ${prev_tag}..${curr_tag} | grep -oP "Co-authored-by: \K.*(?= <)" | sort -u | sed -e 's%^%@%' >> authors_raw.txt
+          LC_ALL=en_US.UTF-8 sort -u authors_raw.txt | grep -v 'dependabot\[bot\]' > authors.txt
+          echo "### Changes in this release:" > changelog.txt
+          cat changes.txt >> changelog.txt
+          echo "" >> changelog.txt
+          echo "Thanks to $(perl -pe 's%\n%, %' authors.txt | sed 's%, $%%')" >> changelog.txt
```
Comment (nasbench) [inline .github/workflows/release.yml:35]:
```suggestion
          echo "" >> changelog.txt
          echo "### Acknowledgement" >> changelog.txt
          echo "Thanks to $(perl -pe 's%\n%, %' authors.txt | sed 's%, $%%') for their contribution to this release" >> changelog.txt
```

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+          echo "" >> changelog.txt
+          echo "" >> changelog.txt
+          echo "#### Which Sigma rule package should I use?" >> changelog.txt
+          echo "A detailed explanation can be found in the [Releases.md](Releases.md) file. If you are new to Sigma, we recommend starting with the \"Core\" ruleset." >> changelog.txt
```
Comment (nasbench) [inline .github/workflows/release.yml:39]: I moved the markdown with some changes to the wiki section. https://github.com/SigmaHQ/sigma/wiki/Sigma-Release-Packages we'll discuss it offline

```suggestion
          echo "A detailed explanation can be found in the release [wiki](https://github.com/SigmaHQ/sigma/wiki/Sigma-Release-Packages) file. If you are new to Sigma, we recommend starting with the \`Core\` ruleset." >> changelog.txt
```

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+          name: Release ${{ github.ref_name }}
+          body_path: changelog.txt
+          token: ${{ secrets.GITHUB_TOKEN }}
+          draft: false
+          prerelease: false
+          files: |
+            sigma-core-${{ github.ref_name }}.zip
+            sigma-core+-${{ github.ref_name }}.zip
+            sigma-core++-${{ github.ref_name }}.zip
+            sigma-et-addon-${{ github.ref_name }}.zip
```
Comment (nasbench) [inline .github/workflows/release.yml:62]:
```suggestion
            sigma-emerging-threats-addon-${{ github.ref_name }}.zip
```

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+          tag_name: ${{ github.ref }}
+          name: Release ${{ github.ref_name }}
+          body_path: changelog.txt
+          token: ${{ secrets.GITHUB_TOKEN }}
+          draft: false
+          prerelease: false
+          files: |
+            sigma-core-${{ github.ref_name }}.zip
+            sigma-core+-${{ github.ref_name }}.zip
+            sigma-core++-${{ github.ref_name }}.zip
```
Comment (nasbench) [inline .github/workflows/release.yml:61]: Perhaps we should use underscore to not confuse with the `-` sign from the release?

File: tests/sigma-package-release.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+    print("[I] Parsing and selecting rules, this will take some time...")
+    selected_rules = select_rules(args)
+    print("[I] Selected {} rules".format(len(selected_rules)))
+
+    write_zip(args.outfile, selected_rules)
+    print("[I] Written all rules to output ZIP file '{}'".format(args.outfile))
+
+if __name__ == '__main__':
+    sys.exit(main(sys.argv[1:]))
```
Comment (nasbench) [inline tests/sigma-package-release.py:115]:
```suggestion
    sys.exit(main(sys.argv[1:]))

```

File: .github/workflows/release.yml (hunk 6 above)
Comment (phantinuss) [reply .github/workflows/release.yml:62]: Needs to be fixed in two locations, will do

File: .github/workflows/release.yml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+          echo "" >> changelog.txt
+          echo "#### Which Sigma rule package should I use?" >> changelog.txt
+          echo "A detailed explanation can be found in the [Releases.md](Releases.md) file. If you are new to Sigma, we recommend starting with the \"Core\" ruleset." >> changelog.txt
+          cat changelog.txt
+      - name: Build all release packages
+        run: |
+          python3 tests/sigma-package-release.py --min-status test --min-level high --rule-types generic --outfile sigma-core-${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status test --min-level medium --rule-types generic --outfile sigma-core+-${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level medium --rule-types generic --outfile sigma-core++-${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level high --rule-types et --outfile sigma-et-addon-${{ github.ref_name }}.zip
```
Comment (phantinuss) [inline .github/workflows/release.yml:46]:
```suggestion
          python3 tests/sigma-package-release.py --min-status experimental --min-level high --rule-types et --outfile sigma-emerging-threats-addon-${{ github.ref_name }}.zip
```

File: .github/workflows/release.yml (hunk 7 above)
Comment (phantinuss) [reply .github/workflows/release.yml:61]: Do you mean
`sigma_core++_${{ github.ref_name }}.zip`
or
`sigma_core++-${{ github.ref_name }}.zip`
?

Comment (phantinuss) [reply .github/workflows/release.yml:61]: The release tag doesn't use a `-`. It is currently planned as `r<iso date>` e.g. `r2023-10-06`.

Comment (phantinuss) [reply .github/workflows/release.yml:61]: ok now I see dashes :D

Comment (phantinuss) [reply .github/workflows/release.yml:61]: I mean it's very subjective how it "feels" in the release page. I'll do a test with `_`s to get a direct comparison.

Comment (phantinuss) [reply .github/workflows/release.yml]: dashes
image
vs
underscores
image

Yes, I also prefer the underscores

File: .github/workflows/release.yml [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+          echo "### Which Sigma rule package should I use?" >> changelog.txt
+          echo "A detailed explanation can be found in the [Releases.md](Releases.md) file. If you are new to Sigma, we recommend starting with the \"Core\" ruleset." >> changelog.txt
+          cat changelog.txt
+      - name: Build all release packages
+        run: |
+          python3 tests/sigma-package-release.py --min-status test --min-level high --rule-types generic --outfile sigma_core_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status test --min-level medium --rule-types generic --outfile sigma_core+_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level medium --rule-types generic --outfile sigma_core++_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level high --rule-types et --outfile sigma_emerging_threats_addon_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level high --rule-types generic et --outfile sigma_all_rules_${{ github.ref_name }}.zip
```
Comment (pH-T) [inline .github/workflows/release.yml:47]: --min-level high?

Comment (phantinuss) [reply .github/workflows/release.yml:47]: good catch, thanks

Comment (nasbench) [reply .github/workflows/release.yml:47]: Yes. Nice catch

Comment (phantinuss) [inline .github/workflows/release.yml:47]:
```suggestion
          python3 tests/sigma-package-release.py --min-status experimental --min-level medium --rule-types generic et --outfile sigma_all_rules_${{ github.ref_name }}.zip
```

File: .github/workflows/release.yml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          echo "" >> changelog.txt
+          echo "### Which Sigma rule package should I use?" >> changelog.txt
+          echo "A detailed explanation can be found in the [Releases.md](Releases.md) file. If you are new to Sigma, we recommend starting with the \"Core\" ruleset." >> changelog.txt
+          cat changelog.txt
+      - name: Build all release packages
+        run: |
+          python3 tests/sigma-package-release.py --min-status test --min-level high --rule-types generic --outfile sigma_core_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status test --min-level medium --rule-types generic --outfile sigma_core+_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level medium --rule-types generic --outfile sigma_core++_${{ github.ref_name }}.zip
+          python3 tests/sigma-package-release.py --min-status experimental --min-level high --rule-types et --outfile sigma_emerging_threats_addon_${{ github.ref_name }}.zip
```
Comment (phantinuss) [inline .github/workflows/release.yml:46]:
```suggestion
          python3 tests/sigma-package-release.py --min-status experimental --min-level medium --rule-types et --outfile sigma_emerging_threats_addon_${{ github.ref_name }}.zip
```

**Review by nasbench — APPROVED** [review]

**Pull request merged**